### ISSUE TITLE
[feature request] freeze() for nn.Module

### ISSUE THREAD

**Issue by samuelbroscheit** [pytorch/pytorch#9046]
For Modules like BatchNorm it is not sufficient to set requires_grad = False, because the running_mean and running_var are still updated unless the the module is set to .eval() . 

Unless I am missing something, ATM, I only can solve this with

```
            self.bn.train(self.bn.training and self.bn.weight.requires_grad)
            x = self.bn(x)
```
which feels quite hacky. Also, in general, it would be nice to have a convience function to freeze the parameters of a Module. 

The idea would be that freeze() sets all the parameters of a module to requires_grad = False and sets a flag self.freeze =  True which blocks all future .train() calls to the module.

**Comment by Evpok** [comment]
It's been almost two years, but there is [torch._C_.freeze_module](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/freeze_module.h) which seems to do something like this and [many questions asking for this](https://discuss.pytorch.org/search?q=freeze). Any plans to go forward with this?

**Comment by albanD** [comment]
Note that the functions in `torch._C` are not for public use. And this one in particular is special to the jit.

**Comment by ezyang** [comment]
BTW, the JIT pass is completely unrelated to what this question is asking for; it is involved with freezing other things (not what you're asking for here.)

**Comment by Evpok** [comment]
@ezyang oh, sorry, that's what I understood from skimming the description, but I did not research this deep since indeed these are not meant for public use

**Comment by gchanan** [comment]
we don't think this necessitates a new API (yet) -- I'm going to file a separate issue for updating the documentation to make it clear what the different behavior is for the combination of `requires_grad_` and/or `train`.  We would need to document that anyway, because the behavior isn't obvious from the name, so starting with the documentation seems like the right choice.
